---
title: "P337: Broad term enrichment in modules"
subtitle: "Hypergeometric enrichment and GSEA"
author: "Kim Dill-McFarland, kadm@uw.edu"
output:
  html_document:
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
date: "version `r format(Sys.time(), '%B %d, %Y')`"
editor_options: 
  chunk_output_type: console
---

# Background

The purpose of this workflow is to determine gene module functions through hypergeometric enrichment of Broad gene set genes in modules and gene set enrichment analysis (GSEA) of fold changes in gene expression (V5 - V4).

# Setup
Load packages

```{r message=FALSE, warning=FALSE}
# Data manipulation and figures
library(tidyverse)
#Genome
library(org.Hs.eg.db)
# Print tty table to knit file
library(knitr)
library(kableExtra)
options(knitr.kable.NA = '')
```

Set seed

```{r}
set.seed(4389)
```

Scripts 

```{r}
#Script hypergeo enrichment
source("https://raw.githubusercontent.com/kdillmcfarland/R_bioinformatic_scripts/master/hypergeo_enricher.R")
#Script for running GSEA
source("https://raw.githubusercontent.com/kdillmcfarland/R_bioinformatic_scripts/master/GSEA_fxn.R")
```

# Load data

Lists of genes in modules.

```{r}
#Read in all genes in module lists
net.files <- list.files(path="results/module_level/", pattern="genes_in_mod.csv",
                        recursive = TRUE, full.names = TRUE)

net.all <- data.frame()
for(file in net.files){
  net.temp <- read_csv(file) %>% 
    mutate(file = basename(file))
  
  net.all <- rbind(net.all, net.temp)
}

net.all <- net.all %>% 
  #remove module 00
  filter(module.char != "00") %>% 
  mutate(file = gsub("_genes_in_mod.csv", "", file),
         file = gsub("P337_", "", file),
         module = paste(file, module.char, sep="_")) %>% 
  dplyr::select(module, geneName, hgnc_symbol)
```

Fold change.

```{r}
#Gene fold changes
## automatically calculated by limma
FC <- read_csv("results/gene_level/P337_BAL_gene_visit.csv") %>% 
  mutate(cell="BAL") %>% 
  bind_rows(mutate(read_csv("results/gene_level/P337_BAL_gene_visit.csv"), cell="BE")) %>% 
  filter(group == "visit") %>% 
  dplyr::select(cell, geneName, logFC)
```

# Gene set descriptions

Gene ontology ([GO](http://geneontology.org/)) provides gene functions and annotations from a variety of sources. Specifics on the gene sets below can be found at <http://software.broadinstitute.org/gsea/msigdb/collections.jsp>.

* Hallmark gene sets (H)
    - Hallmark gene sets summarize and represent specific well-defined biological states or processes and display coherent expression. These gene sets were generated by a computational methodology based on identifying overlaps between gene sets in other MSigDB collections and retaining genes that display coordinate expression. 
* Curated gene sets (C2)
    - Gene sets curated from various sources including online pathway databases, the biomedical literature, and knowledge of domain experts. Includes:
    - CGP: chemical and genetic perturbations
    - CP: Canonical pathways
        * BIOCARTA
        * KEGG
        * PID
        * REACTOME
* Ontology gene sets (C5)
    - Gene sets that contain genes annotated by the same ontology term. The C5 collection is divided into two sub-collections, the first derived from the Gene Ontology resource (GO) which contains BP, CC, and MF components and a second derived from the Human Phenotype Ontology (HPO). 
        * BP: Biological process
        * CC: Cellular component
        * MF: Molecular function
    
# Hypergeometric enrichment

Determine if genes in a set are enriched in modules.

## Hallmark (H)

```{r eval=FALSE}
dir.create("results/enrichment/", showWarnings = FALSE)

enrich.fxn(gene.df = net.all, 
           df.group="module",
           category = "H",
           ID.type = "ENSEMBL",
           genome = org.Hs.eg.db, 
           basename = "modules",
           outdir = "results/enrichment/")
```

Significant enrichments, FDR $\leq$ 0.2.

```{r warning=FALSE, echo=FALSE, message=FALSE}
H <- read_csv("results/enrichment/enrich_modules_H.csv")
  
H %>% 
  filter(p.adjust<=0.2) %>% 
  dplyr::select(group, size.group,
    Description, size.overlap.term, size.term, 
    p.adjust) %>% 
  arrange(group, p.adjust) %>% 
  
  kable(col.names = c("Module", "Genes in group",
                      "Description", "Genes overlap",
                      "Genes in term",
                      "FDR")) %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE) %>% 
  collapse_rows(1:2, valign="top") %>% 
  pack_rows("BAL EOS", 1, 32) %>% 
  pack_rows("BAL LYM", 33, 36) %>% 
  pack_rows("BAL MONO", 37, 59) %>% 
  pack_rows("BAL PMN", 60, 67) %>% 
  pack_rows("BAL Unassign", 68, 73)%>% 
  pack_rows("BE", 74, 104)
```

## Canonical pathways (C2 CP)

```{r eval=FALSE}
enrich.fxn(gene.df = net.all, 
           df.group="module",
           category = "C2",
           subcategory = "CP",
           ID.type = "ENSEMBL",
           genome = org.Hs.eg.db, 
           basename = "modules",
           outdir = "results/enrichment/")
```

Significant enrichments, FDR $\leq$ 0.2.

```{r warning=FALSE, echo=FALSE, message=FALSE}
CP <- read_csv("results/enrichment/enrich_modules_C2_CP.csv")
  
CP %>% 
  filter(p.adjust<=0.2) %>% 
  dplyr::select(group, size.group,
    Description, size.overlap.term, size.term, 
    p.adjust) %>% 
  arrange(group, p.adjust) %>% 
  
  kable(col.names = c("Module", "Genes in group",
                      "Description", "Genes overlap",
                      "Genes in term",
                      "FDR")) %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE) %>% 
  collapse_rows(1:2, valign="top") %>% 
  pack_rows("BAL EOS", 1,17) %>% 
  pack_rows("BAL LYM", 18,18) %>% 
  pack_rows("BAL MONO", 19,31) %>% 
  pack_rows("BAL Unassign", 32,49) %>% 
  pack_rows("BE", 50,63)
```

## Ontology (C5)

```{r eval=FALSE}
enrich.fxn(gene.df = net.all, 
           df.group="module",
           category = "C5",
           ID.type = "ENSEMBL",
           genome = org.Hs.eg.db, 
           basename = "modules",
           outdir = "results/enrichment/")
```

Significant enrichments, minimum 3 FDR values that are < 0.2 for each.

```{r warning=FALSE, echo=FALSE, message=FALSE}
C5 <- read_csv("results/enrichment/enrich_modules_C5.csv")
  
C5 %>% 
  filter(p.adjust <= 0.2) %>% 
  group_by(group) %>% 
  slice_min(order_by = p.adjust, n=3) %>% 
  filter(!grepl("_00", group)) %>% 
  dplyr::select(group, size.group,
    Description, size.overlap.term, size.term, 
    p.adjust) %>% 
  arrange(group, p.adjust) %>% 
  
  kable(col.names = c("Module", "Genes in group",
                      "Description", "Genes overlap",
                      "Genes in term",
                      "FDR")) %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE) %>% 
  collapse_rows(1:2, valign="top") %>% 
  pack_rows("BAL EOS", 1, 31) %>% 
  pack_rows("BAL LYM", 32, 36) %>% 
  pack_rows("BAL MONO", 37, 65) %>% 
  pack_rows("BAL PMN", 66, 68) %>% 
  pack_rows("BAL Unassign", 69, 89) %>%  
  pack_rows("BE", 90, 123)
```

# Gene set enrichment analysis (GSEA)

Prepare fold change gene lists

```{r message=FALSE}
gene.ls <- list()
for(mod in unique(net.all$module)){
  genes.temp <- net.all %>% 
    filter(module == mod) %>% 
    dplyr::select(geneName, hgnc_symbol) %>% 
    left_join(FC, by = "geneName") %>% 
    distinct(hgnc_symbol, logFC) %>% 
    #Merge duplicate HGNC
    group_by(hgnc_symbol) %>% 
    dplyr::summarise(logFC = mean(logFC, na.rm=TRUE), .groups="drop")
  
  #Save to list
  gene.ls[[mod]] <- genes.temp$logFC
    names(gene.ls[[mod]]) <- genes.temp$hgnc_symbol
}
```

## Hallmark (H)

```{r eval=FALSE}
GSEA(gene_list = gene.ls, 
     gmt_file="~/Documents/refs/Broad_gmt/h.all.v7.1.symbols.gmt",
     outdir="results/enrichment/", plotdir="figs/enrichment/", 
      plot=TRUE, plot.fdr=0.1)
```

Modules and gene sets with at least 1 FDR < 0.1

```{r echo=FALSE, message=FALSE, fig.width=8.5}
H_GSEA <- read_csv("results/enrichment/h_GSEA.result.csv")

#Filter to significant results
to.plot <- H_GSEA %>% 
    filter(fgsea.FDR <= 0.1)
  
plot.dat <- H_GSEA %>% 
    #Significant terms
    filter(pathway %in% to.plot$pathway & group %in% to.plot$group) %>% 
    #color by significance
    mutate(Significance = ifelse(fgsea.FDR <= 0.01, "FDR < 0.01",
                          ifelse(fgsea.FDR <= 0.05, "FDR < 0.05",
                          ifelse(fgsea.FDR <= 0.1, "FDR < 0.1",
                          ifelse(fgsea.FDR <= 0.2, "FDR < 0.2",
                                 "NS"))))) %>% 
  #Shorten labels
  mutate(pathway = gsub("HALLMARK_","",pathway))
  
#Enrichment score limits
plot.lim <- max(abs(plot.dat$fgsea.NES))+0.2
   
plot.dat %>%  
    ggplot(aes(reorder(pathway, fgsea.NES), fgsea.NES)) +
    geom_segment(aes(reorder(pathway, fgsea.NES), 
                      xend=pathway, y=0, yend=fgsea.NES)) +
    geom_point(size=3, aes(fill = Significance),
               shape=21, stroke=1) +
    geom_hline(yintercept = 0) +
    
    scale_fill_manual(values=c("FDR < 0.05"="#a50026",
                               "FDR < 0.1"="#ffffbf",
                               "NS"="grey")) +
    lims(y=c(-plot.lim,plot.lim)) +
    coord_flip() +
    labs(x="Pathway", y="Normalized Enrichment Score",
         title="Hallmark GSEA", fill = "FGSEA significance") + 
    facet_grid( ~ group) +
    theme_bw() +
    theme(legend.position = "bottom")
```

# R session

```{r}
sessionInfo()
```

***
